aot/src/main/java/com/dylibso/chicory/aot/AotAnalyzer.java

```diff
@@ -0,0 +1,731 @@
+import static com.dylibso.chicory.aot.TypeStack.FUNCTION_SCOPE;
+import static java.util.Collections.reverse;
+import static java.util.stream.Collectors.toCollection;
+import static java.util.stream.Collectors.toUnmodifiableList;
+
+import com.dylibso.chicory.wasm.Module;
+import com.dylibso.chicory.wasm.exceptions.ChicoryException;
+import com.dylibso.chicory.wasm.types.AnnotatedInstruction;
+import com.dylibso.chicory.wasm.types.ExternalType;
+import com.dylibso.chicory.wasm.types.FunctionBody;
+import com.dylibso.chicory.wasm.types.FunctionImport;
+import com.dylibso.chicory.wasm.types.FunctionType;
+import com.dylibso.chicory.wasm.types.Global;
+import com.dylibso.chicory.wasm.types.GlobalImport;
+import com.dylibso.chicory.wasm.types.Instruction;
+import com.dylibso.chicory.wasm.types.OpCode;
+import com.dylibso.chicory.wasm.types.Table;
+import com.dylibso.chicory.wasm.types.TableImport;
+import com.dylibso.chicory.wasm.types.ValueType;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.IntStream;
+import java.util.stream.LongStream;
+import java.util.stream.Stream;
+
+final class AotAnalyzer {
+
+    private final Module module;
+    private final List<ValueType> globalTypes;
+    private final List<FunctionType> functionTypes;
+    private final List<ValueType> tableTypes;
+    private final int functionImports;
+
+    public AotAnalyzer(Module module) {
+        this.module = module;
+        this.globalTypes = getGlobalTypes(module);
+        this.functionTypes = getFunctionTypes(module);
+        this.tableTypes = getTableTypes(module);
+        this.functionImports = module.importSection().count(ExternalType.FUNCTION);
+    }
+
+    public List<ValueType> globalTypes() {
+        return globalTypes;
+    }
+
+    public List<FunctionType> functionTypes() {
+        return functionTypes;
+    }
+
+    public List<AotInstruction> analyze(int funcId) {
+        var functionType = functionTypes.get(funcId);
+        var body = module.codeSection().getFunctionBody(funcId - functionImports);
+        var stack = new TypeStack();
+        int nextLabel = body.instructions().size();
+        List<AotInstruction> result = new ArrayList<>();
+
+        // find label targets
+        Set<Integer> labels = new HashSet<>();
+        for (AnnotatedInstruction ins : body.instructions()) {
+            if (ins.labelTrue() != AnnotatedInstruction.UNDEFINED_LABEL) {
+                labels.add(ins.labelTrue());
+            }
+            if (ins.labelFalse() != AnnotatedInstruction.UNDEFINED_LABEL) {
+                labels.add(ins.labelFalse());
+            }
+            labels.addAll(ins.labelTable());
+        }
+
+        // implicit block for the function
+        stack.enterScope(FUNCTION_SCOPE, FunctionType.of(List.of(), functionType.returns()));
```

Comment:
Are we discarding `functionType.params()`?

This code seems to make use of them:
https://github.com/dylibso/chicory/pull/596/files#diff-89b97bafcb107230b3864b1fc861e71f0eef8784e78fed7987356111d89f632cR634

Reply:
Good observation, but this is actually a "block type" rather than a "function type". The parameters for a block are on the stack when the block is entered. The function's implicit block is different, as function arguments are in locals rather than on the stack.